File: typescript/session2/td/heritageMultiple/AlgebreNatParInt.ts
import {Nat} from "../../../session1/td/naturels"

export interface AlgebreNatParInt extends Nat {
    somme(val : Nat) : Nat;
    zero() : Nat;
    produit(x : Nat) : Nat;
    un() : Nat;
    modulo(x : Nat) : Nat
    div(x: Nat) : Nat;
    estEgal(x : Nat) : boolean;
    representation() : string;
}

export abstract class AlgebreNatParInt_defaut implements AlgebreNatParInt {
    abstract creerNatAvecValeur(valeur: number) : Nat;
    abstract creerNatAvecRepresentation(repDecimale : String) : Nat;
    abstract val() : number;
    abstract estNul() : boolean;
    abstract predecesseur() : Nat;
    abstract taille() : number;
    abstract chiffre(i : number) : number;
    abstract creerZero() : Nat;
    abstract creerSuccesseur(predecesseur : Nat) : Nat;

    somme(x : Nat) : Nat {
        return this.creerNatAvecValeur(x.val() + this.val());
    }

    zero() : Nat {
        return this.creerNatAvecValeur(0);
    }

    produit(x : Nat) : Nat {
        return this.creerNatAvecValeur(x.val() * this.val());
    }

    un() : Nat {
        return this.creerNatAvecValeur(1);
    }

    modulo(x : Nat) : Nat {
        return this.creerNatAvecValeur(this.val() % x.val());
    }

    div(x : Nat) : Nat {
        return this.creerNatAvecValeur(Math.floor(this.val() / x.val()));
    }

    estEgal(n : Nat) : boolean {
        return this.representation() === n.representation();
    }

    representation() : string {
        return String(this.val())
    }
}

File: java/cours/src/session2/td/heritageMultiple/AlgebreNatParInt.java
package session2.td.heritageMultiple;

import session1.td.Nat;

public interface AlgebreNatParInt extends Nat{
	

	@Override
	default public Nat zero() {
		return this.creerNatAvecValeur(0);
	}

	@Override
	default public Nat somme(Nat x) {
		return this.creerNatAvecValeur(this.val() + x.val());
	}

	@Override
	default public Nat un() {
		return this.creerNatAvecValeur(1);
	}

	@Override
	default public Nat produit(Nat x) {
		return this.creerNatAvecValeur(this.val() * x.val());
	}

	@Override
	default Nat modulo(Nat x) {
		return this.creerNatAvecValeur(this.val() % x.val());
	}

	@Override
	default Nat div(Nat x) {
		return this.creerNatAvecValeur(this.val() / x.val());
	}

	
	default String representer() {
		return Integer.toString(this.val());
	}
	
	default boolean estEgal(Object obj) {
		if(!(obj instanceof Nat))
			return false;
		Nat x = (Nat)obj;
		return this.val() == x.val();
	}
}
